ta/trusted_key/entry.c

```diff
@@ -0,0 +1,308 @@
+	TEE_CloseTASession(sess);
+
+	return res;
+}
+
+static TEE_Result huk_AE_encrypt(TEE_OperationHandle crypto_op, void *in,
+				 size_t in_sz, void *out, size_t *out_sz)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t *iv = (uint8_t *)out;
+	uint8_t *tag = (uint8_t *)out + IV_SIZE;
+	uint8_t *enc_key = (uint8_t *)out + IV_SIZE + TAG_SIZE;
+	size_t enc_key_len = in_sz, tag_len = TAG_SIZE;
+
+	TEE_GenerateRandom(iv, IV_SIZE);
+
+	res = TEE_AEInit(crypto_op, iv, IV_SIZE, TAG_SIZE * 8, 0, 0);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = TEE_AEEncryptFinal(crypto_op, in, in_sz, enc_key, &enc_key_len,
+				 tag, &tag_len);
+	if (res != TEE_SUCCESS || tag_len != TAG_SIZE) {
+		res = TEE_ERROR_SECURITY;
+		return res;
+	}
+
+	if (ADD_OVERFLOW(enc_key_len, (IV_SIZE + TAG_SIZE), out_sz))
+		return TEE_ERROR_SECURITY;
+
+	return res;
+}
+
+static TEE_Result huk_AE_decrypt(TEE_OperationHandle crypto_op, void *in,
+				 size_t in_sz, void *out, size_t *out_sz)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t *iv = (uint8_t *)in;
+	uint8_t *tag = (uint8_t *)in + IV_SIZE;
+	uint8_t *enc_key = (uint8_t *)in + IV_SIZE + TAG_SIZE;
+	size_t enc_key_len = 0;
+
+	if (SUB_OVERFLOW(in_sz, (IV_SIZE + TAG_SIZE), &enc_key_len))
+		return TEE_ERROR_SECURITY;
+
+	res = TEE_AEInit(crypto_op, iv, IV_SIZE, TAG_SIZE * 8, 0, 0);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = TEE_AEDecryptFinal(crypto_op, enc_key, enc_key_len, out, out_sz,
+				 tag, TAG_SIZE);
+	if (res != TEE_SUCCESS)
+		res = TEE_ERROR_SECURITY;
+
+	return res;
+}
+
+static TEE_Result huk_crypt(TEE_OperationMode mode, void *in, size_t in_sz,
+			    void *out, size_t *out_sz)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_OperationHandle crypto_op = TEE_HANDLE_NULL;
+	TEE_ObjectHandle hkey = TEE_HANDLE_NULL;
+	uint8_t huk_key[TA_DERIVED_KEY_MAX_SIZE] = { 0 };
+	TEE_Attribute attr = { 0 };
+
+	res = TEE_AllocateOperation(&crypto_op, TEE_ALG_AES_GCM, mode,
+				    sizeof(huk_key) * 8);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = derive_unique_key(huk_key, sizeof(huk_key), NULL, 0);
+	if (res != TEE_SUCCESS) {
+		EMSG("derive_unique_key failed: returned 0x%x\n", res);
+		goto out_op;
+	}
+
+	res = TEE_AllocateTransientObject(TEE_TYPE_AES, sizeof(huk_key) * 8,
+					  &hkey);
+	if (res != TEE_SUCCESS)
+		goto out_op;
+
+	attr.attributeID = TEE_ATTR_SECRET_VALUE;
+	attr.content.ref.buffer = huk_key;
+	attr.content.ref.length = sizeof(huk_key);
+
+	res = TEE_PopulateTransientObject(hkey, &attr, 1);
+	if (res != TEE_SUCCESS)
+		goto out_key;
+
+	res = TEE_SetOperationKey(crypto_op, hkey);
+	if (res != TEE_SUCCESS)
+		goto out_key;
+
+	if (mode == TEE_MODE_ENCRYPT) {
+		res = huk_AE_encrypt(crypto_op, in, in_sz, out, out_sz);
+		if (res != TEE_SUCCESS)
+			EMSG("huk_AE_encrypt failed: returned 0x%x\n", res);
+	} else if (mode == TEE_MODE_DECRYPT) {
+		res = huk_AE_decrypt(crypto_op, in, in_sz, out, out_sz);
+		if (res != TEE_SUCCESS)
+			EMSG("huk_AE_decrypt failed: returned 0x%x\n", res);
+	}
```

Comment:
`else { TEE_Panic(0); }`

Reply:
Do we really need to panic here?

Generally the mode of operation of a Linux kernel TEE driver is to open a TA session during TEE device probe, close session during TEE device remove and issue invoke commands on behalf of requests from either user space IOCTLs or other kernel space drivers.

So in case of Linux trusted keys, some improper trusted keys unseal request may come from user space IOCTL to unseal a key. And I think with panic here as decrypt failed would close current session with TA and make the kernel driver useless to handle further requests.

If in case you are worried about large erroneous unseal requests leaking some information regarding HUK, then we can explicitly zeroize output buffer here in case of decrypt failure.

Reply:
Ah, sorry ignore my above comment as now I realized that your comment was regarding outer "if/else". Will add this panic there.